linux/src/threadpool_linux.c

```diff
@@ -0,0 +1,322 @@
+
+#include <stdlib.h>
+#include <inttypes.h>
+#include <stdio.h>
+#include <stdbool.h>
+
+#include "macro_utils/macro_utils.h"
+
+#include "c_logging/xlogging.h"
+
+#include "c_pal/threadapi.h"
+#include "c_pal/interlocked.h"
+#include "c_pal/sync.h"
+#include "c_pal/srw_lock.h"
+#include "c_pal/execution_engine.h"
+#include "c_pal/execution_engine_linux.h"
+
+#include "c_pal/threadpool.h"
+
+#define TIMEOUT_MS      100
+#define MAX_THREAD_COUNT    16
+
+#define POOL_STATE_VALUES \
+    POOL_STATE_UNINIT, \
+    POOL_STATE_ACTIVE, \
+    POOL_STATE_IDLE
+
+MU_DEFINE_ENUM(POOL_STATE, POOL_STATE_VALUES)
+
+typedef struct THREADPOOL_TASK_TAG
+{
+    THREADPOOL_WORK_FUNCTION task_func;
+    void* task_param;
+    struct THREADPOOL_TASK_TAG* next;
+} THREADPOOL_TASK;
+
+typedef struct THREADPOOL_TAG
+{
+    volatile_atomic int32_t thread_count;
+    uint32_t max_thread_count;
+    volatile_atomic int32_t pool_state;
+    volatile_atomic int32_t quitting;
+    volatile_atomic int32_t task_count;
+
+    // Pool Queue
+    THREADPOOL_TASK* task_list;
+    THREADPOOL_TASK* task_list_last;
+
+    SRW_LOCK_HANDLE srw_lock;
+
+    THREAD_HANDLE* thread_handle_list;
+    uint32_t list_index;
+} THREADPOOL;
+
+int threadpool_work_func(void* param)
+{
+    if (param == NULL)
+    {
+        LogCritical("Invalid args: param: %p", param);
+    }
+    else
+    {
+        THREADPOOL* threadpool = param;
+        do
+        {
+            (void)interlocked_exchange(&threadpool->pool_state, POOL_STATE_IDLE);
+            int32_t current_value;
+
+            while ((current_value = interlocked_add(&threadpool->task_count, 0)) == 0 && interlocked_add(&threadpool->quitting, 0) != 1)
+            {
+                if (!wait_on_address(&threadpool->task_count, current_value, TIMEOUT_MS))
+                {
+                    // Timed out, loop through and try again
+                }
+                else
+                {
+                    // There was a new item added to the list, let's go
+                    break;
+                }
+            }
+            (void)interlocked_exchange(&threadpool->pool_state, POOL_STATE_ACTIVE);
+
+            // Locking
+            srw_lock_acquire_exclusive(threadpool->srw_lock);
+            if (threadpool->task_list != NULL && interlocked_add(&threadpool->quitting, 0) != 1)
+            {
+                // Remove the task item from the list
+                THREADPOOL_TASK* temp_item = threadpool->task_list;
+                threadpool->task_list = temp_item->next;
+                (void)interlocked_decrement(&threadpool->task_count);
+
+                srw_lock_release_exclusive(threadpool->srw_lock);
+
+                temp_item->task_func(temp_item->task_param);
+                free(temp_item);
+
+                srw_lock_acquire_exclusive(threadpool->srw_lock);
+            }
+            // Unlocking
+            srw_lock_release_exclusive(threadpool->srw_lock);
```

Comment:
extra acquire/release can be avoided, like

```
acquire
if (task_lst!=NULL...)
{
    temp_item = ...
}
release
if (temp_item != NULL...
```

Reply:
Ah!  I like it, good comment.